status-quo story: Alan needs async in traits (#134)
rust-lang/wg-async-foundations: src/vision/status_quo/alan_needs_async_in_traits.md

# 😱 Status quo stories: Alan needs async in traits

## 🚧 Warning: Draft status 🚧

This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 

If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!

## The story

> Alan is working on a project with Barbara which has already gotten off to a [somewhat rocky start](https://github.com/rust-lang/wg-async-foundations/pull/111). He is working on abstracting away the HTTP implementation the library uses so that users can provide their own. He wants the user to implement an async trait called `HttpClient` which has one method `perform(request: Request) -> Response`. Alan tries to create the async trait:

Reviewer: Let's link the to merged story, not the PR:

```suggestion
Alan is working on a project with Barbara which has already gotten off to a [somewhat rocky start](./barbara_anguishes_over_http.md). He is working on abstracting away the HTTP implementation the library uses so that users can provide their own. He wants the user to implement an async trait called `HttpClient` which has one method `perform(request: Request) -> Response`. Alan tries to create the async trait:
```

```rust,ignore
trait HttpClient {
    async fn perform(request: Request) -> Response;
}
```

When Alan tries to compile this, he gets an error:

```ignore
 --> src/lib.rs:2:5
  |
2 |     async fn perform(request: Request) -> Response;
  |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |     |
  |     `async` because of this
  |
  = note: `async` trait functions are not currently supported
  = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
```

> Alan, using Rust for a little while now, has learned to follow compiler error messages and adds `async-trait` to the dependencies list. Alan follows the README of `async-trait` and gets the code to compile. Alan is somewhat dismayed that the compile times have increased considerably after adding this new dependency.

Reviewer 2: ```suggestion
Alan, who has been using Rust for a little while now, has learned to follow compiler error messages and adds `async-trait` to his `Cargo.toml`. Alan follows the README of `async-trait`, and comes up with the following code:
'''rust
#[async_trait]
trait HttpClient {
    async fn perform(request: Request) -> Response;
}
'''
Despite getting his code to compile, Alan is somewhat dismayed that the compile times have increased considerably after adding this new dependency.
```
(markdown doesn't like nested code blocks so I used apostrophes instead of backticks - don't click "apply suggestion" :)

Reviewer 3: +1 for having a code example.

Reviewer 3: Maybe include @rylev's numbers:

---

```suggestion
Alan, who has been using Rust for a little while now, has learned to follow compiler error messages and adds `async-trait` to his `Cargo.toml`. Alan follows the README of `async-trait` and comes up with the following code:
'''rust
#[async_trait]
trait HttpClient {
    async fn perform(request: Request) -> Response;
}
'''
Alan's code now compiles, but he also finds that his compile times have gone from under a second to around 6s, at least for a clean build.
```

> **TODO::** Maybe add representive pre and post trait compile times here.

Reviewer 4: I think a small and simple mention like "the compile times doubled" or "went from 10s to 20s" is good

Reviewer 5: FWIW adding `async-trait` to an empty project brings clean build times from ~0.4s to ~5.75s on my machine.

Author: and that's too big an impact, for it to be a non-concern.

Reviewer 3: Can we include these numbers? That's very interesting.

Reviewer 3: I took a stab at that above :)

Author: Thanks @nikomatsakis.

> After Alan finishes adding the new trait, he shows his work off to Barbara and mentions he's happy with the work but is a little sad that compile times have worsened. Barbara, an experienced Rust developer, knows that using `async-trait` comes with some issues and in this case is especially worried about tying their public API to a third-party dependency. She decides to not worry Alan with this right now. Alan and Barbara are pretty happy with the results and go on to publish their crate which gets lots of users.

Reviewer 5: ```suggestion
After Alan finishes adding the new trait, he shows his work off to Barbara and mentions he's happy with the work but is a little sad that compile times have worsened. Barbara, an experienced Rust developer, knows that using `async-trait` comes with some additional issues. In this particular case she is especially worried about tying their public API to a third-party dependency. Even though it is technically possible to implement traits annotated with `async_trait`, doing so in practice is very painful. She decides to not worry Alan with this right now. Alan and Barbara are pretty happy with the results and go on to publish their crate which gets lots of users.
```

Author: > Even though it is technically possible to implement traits annotated with `async_trait`, doing so in practice is very painful

We should describe at least one example of a pain point here IMO (IOW need to be a bit more specific).

Reviewer 5: I think some pain points are: 
* `async_trait` handles lifetimes for you if the returned future is tied to the lifetime of some inputs
* `async_trait` also boxes and pins the futures for you.

Reviewer 3: I agree it would be nice to have some examples, or a link to them -- I was very curious what this meant.

Author: @nikomatsakis I added the pain points @rylev listed above. Do we need more?

Reviewer 3: I don't know of any more!

Later on, a potential user of the library wants to use their library in a `no_std` context where they will be providing a custom HTTP stack. Alan and Barbara have done a pretty good job of limiting the use of standard library features and think it might be possible to support this use case. However, they quickly run into a show stopper: `async-trait` boxes all of the futures returned from a async trait function. They report this to Alan through an issue.

Alan, feeling (over-) confident in his Rust skills, decides to try to see if he can implement async traits without using `async-trait`. 

```rust 
struct Request(String);

trait HttpClient {

>    type Response: Future<Output = ()>;

Reviewer 5: ```suggestion
   type Response: Future<Output = Response>;
```

Author: @rylev are you sure about this? `Response` associated type is a `Future` that gives `Response` type?

Reviewer 5: Yes, we can change the name of the associated type, but the future will eventually yield an HTTP response

Reviewer 2: Yeah, that is the equivalent to `async fn perform(request: Request) -> Response;`

Author: @ibraheemdev thanks, yeah I know but just got confused because @rylev chose the same name for the output type as the associated type but he explained so all good. :)

   fn perform(request: Request) -> Self::Response; 
}
```

Alan seems to have something working, but when he goes to update the examples of how to implement this trait in his crate's documentation, he realizes that he either needs to:

* use trait object:

>   ```rust,ignore

Reviewer 5: nit: I don't think ignore is necessary here since none of the code samples are run.

Author: Ah, I saw `ignore` being used in other parts, so I added to be sure. :)

Reviewer 3: It used to be require, but we stopped running `mdbook test`

  struct ClientImpl;

  impl HttpClient for ClientImpl {
      type Response = Pin<Box<dyn Future<Output = ()>>>;

      fn perform(request: Request) -> Self::Response {
          Box::pin(async move {
              println!("{}", request.0);
          })
	  }
  }
  ```

  which wouldn't work for `no_std`.

* implement `Future` trait manually, which isn't particulary easy/straight-forward for non-trivial cases, especially if it involves making other async calls (likely).

> **TODO::** Would make sense for Alan to reach out to Barbara at this point and she should be able to give some good advice.

Reviewer 3: There isn't *that* much he can do, right? Maybe I don't know the tricks =)

Reviewer 4: She might suggest writing a custom proc-macro that is like async-trait, but has a special no_std mode which allows using it without std? Alan as a newbie in Rust doesn't know much about proc-macros and how to write them himselfes. It's also a huge project for a Rust beginner, but for Barbara as an advanced Rust Developer, she probably has experience with proc-macros and can help out to write it.

Another tip she might give, is how to implement the `Future` trait properly and how to use / create helpers like the `std::task::ready` macro and other stuff.

Reviewer 2: There is the `poll_x` pattern, and I saw a [cool crate recently that can help with that](https://crates.io/crates/async-trait-ext).... but I guess that falls under implementing the `Future` manually/with a different proc-macro.

Reviewer 3: The only way to make things work with no-std that I know of is to rely on nightly only features, is the thing.

Reviewer 5: I would simply have Barbara mention exactly what Niko said - it's likely not possible without using nightly-only features. We can then have compiling on stable as a part of their requirements, leading this to be a dead end.

Author: What I was hoping for here is that some of you might know of some API that **somehow** allows you to turn an `impl Future` into a named future, similar to how [`futures::stream::unfold`](https://docs.rs/futures/0.3.13/futures/stream/fn.unfold.html) allows you to create a stream out of an `impl Future` but I forgot that will [require boxing](https://gitlab.freedesktop.org/dbus/zbus/-/blob/491e5456d66510110ee8e730f3bb9e5a692844c0/zbus/src/azync/proxy.rs#L310) as well.

As for a proc-macro way, from what I understood from the [README](https://crates.io/crates/async-trait-ext), `async-trait-ext` only requires a nightly feature for provided methods in traits. Alan doesn't necessarily need provided methods so he can possibly make use of `async-trait-ext` w/o nightly, can't he?

> 

Reviewer 5: Let's add a final sentence to wrap this up. Something like "Alan and Barbara accept that they won't be able to support `no_std` users of their library."

Author: thinking more about it, it's actually not `no_std` that's impossible here but rather `no_alloc`, although I'm not sure if this typically makes any difference in the real world.

## 🤔 Frequently Asked Questions

### **What are the morals of the story?**

* `async-trait` is awesome, but has some drawbacks

>     * compile time increases

Reviewer 2: Is the compile time increase really that bad?

Reviewer 3: I too wondered about this...

Reviewer: I did a quick test with two dummy projects.

For the code using async-trait, it looks like this:

```rust
#[async_trait]
trait TraitA {
    async fn A();
}

#[async_trait]
impl TraitA for () {
    async fn A() {
        todo!()
    }
}
```

For the code without async-trait, it looks like:

```rust
trait TraitA {
    type R: Future<Output = ()>;
    fn A() -> Self::R;
}

impl TraitA for () {
    type R = Pin<Box<dyn Future<Output = ()>>>;
    fn A() -> Self::R {
        Box::pin(async move {
            todo!()
        })
    }
}
```

I duplicated these traits a total of 7 times (so I have `traitA` through `traitF`)

Using a build script that simply touches main.rs, hyperfine says:

```
$ hyperfine ./foo-async-trait/build.sh ./foo/build.sh
Benchmark #1: ./foo-async-trait/build.sh
  Time (mean ± σ):     609.0 ms ±  11.9 ms    [User: 500.4 ms, System: 125.1 ms]
  Range (min … max):   591.9 ms … 623.3 ms    10 runs

Benchmark #2: ./foo/build.sh
  Time (mean ± σ):     505.7 ms ±   5.1 ms    [User: 398.4 ms, System: 122.5 ms]
  Range (min … max):   499.2 ms … 514.5 ms    10 runs

Summary
  './foo/build.sh' ran
    1.20 ± 0.03 times faster than './foo-async-trait/build.sh'
```

Reviewer 2: That doesn't seem like *that* big of a deal. It's also worth noting that many crates will already have `async-trait`'s deps (`proc-macro`, `syn`, `quote`) as a transitive dep from elsewhere.

Reviewer 5: Did you test clean build times? Considering that this would likely not impact build times _at all_ if it were a built in language feature, I think having build times impacted at all is a concern that should be taken into account.

    * performance cost of boxing and dynamic dispatch

>     * lack of control over whether futures are `Send` or `Sync`

Reviewer 2: `async-trait` *does* give you control over this. You can use `#[async_trait(?Send)]` if you don't want to add `Send` bounds.

Reviewer 3: I'd say we should remove this bullet

    * not a standard solution so when this comes to language, it might break things
* Trying to have a more efficient implementation than `async-trait` is likely not possible. 

### **What are the sources for this story?**

* [Zeeshan](https://github.com/zeenix/) is looking for a way to implement async version of the [service-side zbus API](https://docs.rs/zbus/1.9.1/zbus/trait.Interface.html).
* [Ryan](https://github.com/rylev) had to use `async-trait` in an internal project.

> ## **Why did you choose Alan to tell this story?**

Reviewer 4: There are only two #

Reviewer 3: ```suggestion
### **Why did you choose Alan to tell this story?**
```

```

Alan, who has been using Rust for a little while now, has learned to follow compiler error messages and adds `async-trait` to his `Cargo.toml`. Alan follows the README of `async-trait` and comes up with the following code:

```rust
#[async_trait]
trait HttpClient {
    async fn perform(request: Request) -> Response;
}
```

Alan's code now compiles, but he also finds that his compile times have gone from under a second to around 6s, at least for a clean build.

After Alan finishes adding the new trait, he shows his work off to Barbara and mentions he's happy with the work but is a little sad that compile times have worsened. Barbara, an experienced Rust developer, knows that using `async-trait` comes with some additional issues. In this particular case she is especially worried about tying their public API to a third-party dependency. Even though it is technically possible to implement traits annotated with `async_trait` without using `async_trait`, doing so in practice is very painful. For example `async_trait`:

* handles lifetimes for you if the returned future is tied to the lifetime of some inputs.
* boxes and pins the futures for you.

> which implementer will have to manually handle if they don't use `async_trait`. She decides to not worry Alan with this right now. Alan and Barbara are pretty happy with the results and go on to publish their crate which gets lots of users.

Reviewer 5: ```suggestion
which the implementer will have to manually handle if they don't use `async_trait`. She decides to not worry Alan with this right now. Alan and Barbara are pretty happy with the results and go on to publish their crate which gets lots of users.
```
